### PR TITLE
[pytorch] [build] [E3] PyTorch 1.11.0

### PR DESCRIPTION
**Pull request by ydaiming**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 1]
```diff
@@ -227,6 +229,42 @@ def test_pytorch_s3_plugin_cpu(pytorch_training, ec2_connection, cpu_only, ec2_i
...
+def test_pytorch_training_torchaudio_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type, pt111_and_above_only):
+    if test_utils.is_image_incompatible_with_instance_type(pytorch_training, ec2_instance_type):
+        pytest.skip(f"Image {pytorch_training} is incompatible with instance type {ec2_instance_type}")
+    execute_ec2_training_test(ec2_connection, pytorch_training, PT_TORCHAUDIO_CMD)
+
+
+@pytest.mark.integration("pt_torchaudio_cpu")
+@pytest.mark.model("N/A")
+@pytest.mark.parametrize("ec2_instance_type", PT_EC2_CPU_INSTANCE_TYPE, indirect=True)
+def test_pytorch_traning_torchaudio_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type, pt111_and_above_only):
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:244]: Please correct the spelling of training here:
```suggestion
def test_pytorch_training_torchaudio_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type, pt111_and_above_only):
```

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 2]
```diff
@@ -227,6 +229,42 @@ def test_pytorch_s3_plugin_cpu(pytorch_training, ec2_connection, cpu_only, ec2_i
...
+def test_pytorch_training_torchdata_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type, pt111_and_above_only):
+    if test_utils.is_image_incompatible_with_instance_type(pytorch_training, ec2_instance_type):
+        pytest.skip(f"Image {pytorch_training} is incompatible with instance type {ec2_instance_type}")
+    execute_ec2_training_test(ec2_connection, pytorch_training, PT_TORCHDATA_CMD)
+
+
+@pytest.mark.integration("pt_torchdata_cpu")
+@pytest.mark.model("N/A")
+@pytest.mark.parametrize("ec2_instance_type", PT_EC2_CPU_INSTANCE_TYPE, indirect=True)
+def test_pytorch_traning_torchdata_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type, pt111_and_above_only):
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:262]: Please correct the spelling of training here:
```suggestion
def test_pytorch_training_torchdata_cpu(pytorch_training, ec2_connection, cpu_only, ec2_instance_type, pt111_and_above_only):
```

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchaudio [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+HOME_DIR=/test
+BIN_DIR=${HOME_DIR}/bin
+LOG_DIR=${HOME_DIR}/logs
+
+S3_IO_DEV_BRANCH="v0.11.0-rc2"
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testTorchaudio:7]: Why specifically this branch, rather than the version of torchaudio installed on the image?

Also, please rename the script variable names from `S3_IO*` to `TORCHAUDIO*` or something like that.

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+HOME_DIR=/test
+BIN_DIR=${HOME_DIR}/bin
+LOG_DIR=${HOME_DIR}/logs
+
+S3_IO_DEV_BRANCH="s3-datapipes"
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testTorchdata:7]: Please rename all the script variables from `S3_IO*` to `TORCHDATA*`.

File: test/dlc_tests/container_tests/bin/pytorch_tests/testTorchaudio [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+if [ "$?" -eq "0" ]
+then
+    echo "Torchaudio integration tests succeeded."
+else
+    echo "Torchaudio integration tests failed at ${S3_IO_DEV_REPO} ${S3_IO_DEV_BRANCH} ${S3_IO_TEST_SCRIPT}"
+    exit 1
+fi
+
+echo "Torchaudio integration tests succeeded!"
+exit 0
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testTorchaudio:29]: nit: Please add a blank newline at the end of the script.

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+RUN chmod +x /usr/local/bin/dockerd-entrypoint.py
+
+RUN HOME_DIR=/root \
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
```
Comment (arjkesh) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:281]: ARGS don't persist, so you'll need to redefine them in the stage, or create ENV variables that will persist throughout stages

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt  https://aws-dlc-licenses.s3.amazonaws.com/pytorch-1.10/license.txt
```
Comment (arjkesh) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:234]: note: old license file

**Review by waytrue17 — APPROVED** [review]

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+# Removing link for libtinfo.so.6. This change is needed only for ubuntu 20.04-conda, and can be reverted
+# once conda fixes the issue: https://github.com/conda/conda/issues/9680
+RUN rm -rf /opt/conda/lib/libtinfo.so.6
+
+RUN ${CONDA_PREFIX}/bin/conda config --set ssl_verify False \
+ && pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org \
+ && ln -s ${CONDA_PREFIX}/bin/pip /usr/local/bin/pip3
+
+RUN conda install -c pytorch \
+    magma-cuda115==2.6.1 \
```
Comment (saimidu) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:172]: Why does this need to be hardcoded to `2.6.1`?

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+# Removing link for libtinfo.so.6. This change is needed only for ubuntu 20.04-conda, and can be reverted
+# once conda fixes the issue
+RUN rm -rf /opt/conda/lib/libtinfo.so.6
+
+RUN ${CONDA_PREFIX}/bin/conda config --set ssl_verify False \
+ && pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org \
+ && ln -s ${CONDA_PREFIX}/bin/pip /usr/local/bin/pip3
+
+RUN conda install -c pytorch \
+    magma-cuda115==2.6.1 \
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:168]: Why does this need to be hardcoded to `2.6.1`?

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+
+# Configure Open MPI and configure NCCL parameters
+RUN mv $OPEN_MPI_PATH/bin/mpirun $OPEN_MPI_PATH/bin/mpirun.real \
+ && echo '#!/bin/bash' > $OPEN_MPI_PATH/bin/mpirun \
+ && echo 'mpirun.real --allow-run-as-root "$@"' >> $OPEN_MPI_PATH/bin/mpirun \
+ && chmod a+x $OPEN_MPI_PATH/bin/mpirun \
+ && echo "hwloc_base_binding_policy = none" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+ && echo "rmaps_base_mapping_policy = slot" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+ && echo NCCL_DEBUG=INFO >> /etc/nccl.conf \
+ && echo NCCL_SOCKET_IFNAME=^docker0 >> /etc/nccl.conf
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:245]: nit: Not a merge blocker

Does this need to be in the `e3` and `sagemaker` sections? Can we put this in the `common` section instead, as it doesn't seem to depend on the framework installation?

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+ && git clone https://github.com/aws/aws-ofi-nccl.git -b v${BRANCH_OFI} \
+ && cd aws-ofi-nccl \
+ && ./autogen.sh \
+ && ./configure --with-libfabric=/opt/amazon/efa \
+  --with-mpi=/opt/amazon/openmpi \
+  --with-cuda=/usr/local/cuda \
+  --with-nccl=/usr/local --prefix=/usr/local \
+ && make \
+ && make install \
+ && rm -rf /tmp/efa-ofi-nccl
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:260]: nit: Not a merge blocker

Does this need to be in the `e3` and `sagemaker` sections? Can we put this in the `common` section instead, as it doesn't seem to depend on the framework installation?

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+
+# Install OpenSSH for MPI to communicate between containers, allow OpenSSH to talk to containers without asking for confirmation
+RUN apt-get update \
+ && apt-get install -y  --allow-downgrades --allow-change-held-packages --no-install-recommends \
+ && apt-get install -y --no-install-recommends openssh-client openssh-server \
+ && mkdir -p /var/run/sshd \
+ && cat /etc/ssh/ssh_config | grep -v StrictHostKeyChecking > /etc/ssh/ssh_config.new \
+ && echo "    StrictHostKeyChecking no" >> /etc/ssh/ssh_config.new \
+ && mv /etc/ssh/ssh_config.new /etc/ssh/ssh_config \
+ && rm -rf /var/lib/apt/lists/*
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:270]: nit: Not a merge blocker

Does this need to be in the `e3` and `sagemaker` sections? Can we put this in the `common` section instead, as it doesn't seem to depend on the framework installation?

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+
+# Configure OpenSSH so that nodes can communicate with each other
+RUN mkdir -p /var/run/sshd && \
+ sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd
+
+RUN rm -rf /root/.ssh/ && \
+ mkdir -p /root/.ssh/ && \
+ ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
+ cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+ && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
```
Comment (saimidu) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:280]: nit: Not a merge blocker

Does this need to be in the `e3` and `sagemaker` sections? Can we put this in the `common` section instead, as it doesn't seem to depend on the framework installation?

**Review by saimidu — APPROVED** [review]

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu (hunk 8 above)
Comment (ydaiming) [reply pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:172]: I don't think we need. It was previously pinged to be 2.5.2, and recently released a new 2.6.1. We can try to remove the version tag.
